arduino/cores/packagemanager/package_manager.go

```diff
@@ -288,36 +293,112 @@ func (pme *Explorer) ResolveFQBN(fqbn *cores.FQBN) (
 			fmt.Errorf(tr("board %s not found"), fqbn.StringWithoutConfig())
 	}
 
-	buildProperties, err := board.GetBuildProperties(fqbn.Configs)
+	boardBuildProperties, err := board.GetBuildProperties(fqbn.Configs)
 	if err != nil {
 		return targetPackage, platformRelease, board, nil, nil,
 			fmt.Errorf(tr("getting build properties for board %[1]s: %[2]s"), board, err)
 	}
 
 	// Determine the platform used for the build (in case the board refers
 	// to a core contained in another platform)
-	buildPlatformRelease := platformRelease
-	coreParts := strings.Split(buildProperties.Get("build.core"), ":")
-	if len(coreParts) > 1 {
-		referredPackage := coreParts[0]
-		buildPackage := pme.packages[referredPackage]
-		if buildPackage == nil {
-			return targetPackage, platformRelease, board, buildProperties, nil,
-				fmt.Errorf(tr("missing package %[1]s referenced by board %[2]s"), referredPackage, fqbn)
+	core := boardBuildProperties.Get("build.core")
+	referredCore := ""
+	if split := strings.Split(core, ":"); len(split) > 1 {
+		core, referredCore = split[1], split[0]
+	}
+	variant := boardBuildProperties.Get("build.variant")
+	referredVariant := ""
+	if split := strings.Split(variant, ":"); len(split) > 1 {
+		variant, referredVariant = split[1], split[0]
+	}
+	if referredCore != "" && referredVariant != "" && referredCore != referredVariant {
+		return targetPackage, platformRelease, board, nil, nil,
+			fmt.Errorf(tr("'build.core' and 'build.variant' refer to different platforms: %[1]s and %[2]s"), core+":"+referredCore, variant+":"+referredVariant)
+	}
+
+	var referredPlatformRelease *cores.PlatformRelease
+	referredPackageName := referredCore
+	if referredPackageName == "" {
+		referredPackageName = referredVariant
+	}
+	if referredPackageName != "" {
+		referredPackage := pme.packages[referredPackageName]
+		if referredPackage == nil {
+			return targetPackage, platformRelease, board, nil, nil,
+				fmt.Errorf(tr("missing package %[1]s referenced by board %[2]s"), referredPackageName, fqbn)
+		}
+		referredPlatform := referredPackage.Platforms[fqbn.PlatformArch]
+		if referredPlatform == nil {
+			return targetPackage, platformRelease, board, nil, nil,
+				fmt.Errorf(tr("missing platform %[1]s:%[2]s referenced by board %[3]s"), referredPackageName, fqbn.PlatformArch, fqbn)
+		}
+		referredPlatformRelease = pme.GetInstalledPlatformRelease(referredPlatform)
+		if referredPlatformRelease == nil {
+			return targetPackage, platformRelease, board, nil, nil,
+				fmt.Errorf(tr("missing platform release %[1]s:%[2]s referenced by board %[3]s"), referredPackageName, fqbn.PlatformArch, fqbn)
 		}
-		buildPlatform := buildPackage.Platforms[fqbn.PlatformArch]
-		if buildPlatform == nil {
-			return targetPackage, platformRelease, board, buildProperties, nil,
-				fmt.Errorf(tr("missing platform %[1]s:%[2]s referenced by board %[3]s"), referredPackage, fqbn.PlatformArch, fqbn)
+	}
```

Comment:
Can we make a function of these lines? The block is very wordy

Reply:
ok it should be much cleaner now!